stellargraph/layer/watch_your_step.py

```diff
@@ -0,0 +1,190 @@
+            regularizer=self.attention_regularizer,
+            constraint=self.attention_constraint,
+        )
+
+        self.built = True
+
+    build.__doc__ = Layer.build.__doc__
+
+    def call(self, partial_powers):
+        """
+        Applies the layer and calculates the expected random walks.
+
+        Args:
+            partial_powers: num_rows rows of the first num_powers powers of adjacency matrix with shape
+            (num_rows, num_powers, num_nodes)
+
+        Returns:
+            Tensor that represents the expected random walks starting from nodes corresponding to the input rows of
+            shape (num_rows, num_nodes)
+        """
+
+        attention = K.softmax(self.attention_weights)
+        expected_walk = tf.einsum("ijk,j->ik", partial_powers, attention)
+
+        return expected_walk
+
+
+@experimental(reason="lack of unit tests")
+class WatchYourStep:
+    """
+    Implementation of the node embeddings as in Watch Your Step: Learning Node Embeddings via Graph Attention
+    https://arxiv.org/pdf/1710.09599.pdf.
+
+    This model requires specification of the number of random walks starting from each node, and the embedding dimension
+    to use for the node embeddings. Note, that the embedding dimension should be an even number or else will be
+    rounded down to nearest even number.
+
+    Args:
+        generator (AdjacencyPowerGenerator): the generator
+        num_walks (int): the number of random walks starting at each node to use when calculating the expected random
+            walks. Defaults to `80` as this value was found to perform well by the authors of the paper.
+        embedding dimension (int): the dimension to use for the node embeddings. Defaults to `64`
+            as this value was found to perform well by the authors of the paper.
+        attention_initializer (str or func, optional): The initialiser to use for the attention weights.
+        attention_regularizer (str or func, optional): The regulariser to use for the attention weights.
+        attention_constraint (str or func, optional): The constraint to use for the attention weights.
+    """
+
+    def __init__(
+        self,
+        generator,
+        num_walks=80,
+        embedding_dimension=64,
+        attention_initializer="glorot_uniform",
+        attention_regularizer=None,
+        attention_constraint=None,
+    ):
+
+        if not isinstance(generator, AdjacencyPowerGenerator):
+            raise TypeError(
+                "generator should be an instance of AdjacencyPowerGenerator."
+            )
+
+        if not isinstance(num_walks, int):
+            raise TypeError("num_walks should be an int.")
+
+        if num_walks <= 0:
+            raise ValueError("num_walks should be a positive int.")
+
+        self.num_walks = num_walks
+        self.num_powers = generator.num_powers
+        self.n_nodes = int(generator.Aadj_T.shape[0])
+
+        if embedding_dimension % 2:
```

Comment:
I personally find treating ints as booleans a bit confusing (it takes a little bit to translate the implicit `bool` call to "checking is odd") and would prefer to be explicit:

```suggestion
        if embedding_dimension % 2 != 0:
```